create example-react-hooks-useReducer (#170)
testing-library/testing-library-docs: docs/example-react-hooks-useReducer.md

Basic example showing how to test useReducer hook. 

> The component: changes stateprop1 from false to true which changes some text. 

Reviewer: I'd suggest mentioning that the component changes the DOM, since this is the main benefit of using Testing Library tools - focusing on DOM changes and events, rather than component instances.

```suggestion
The component: changes some text depending on `stateprop1` value.
```

```

> import React, { useReducer } from 'react';

Reviewer: As far as I can tell, most examples in Testing Library docs aren't using the trailing comma, so I'd suggest sticking to it :)

Reviewer 2: Prettier _should_ format the docs on commit, I'm assuming the author wrote this on GH instead of locally?

> import * as ACTIONS from '../store/actions'

Reviewer 2: Can we simplify this down to just one file, with no special constants or anything? Just `dispatch({ type: 'success'} )`. It helps to have the bare minimum in these examples.

import * as Reducer from '../store/reducer'

const TestHookReducer = (props) => {
  const [reducerState, dispatch] = useReducer(Reducer.Reducer1, Reducer.initialState)

  const dispatchActionSuccess = () => {
    dispatch(ACTIONS.SUCCESS)
  }

  const dispatchActionFailure = () => {
    dispatch(ACTIONS.FAILURE)
  }

  return (
    <div>
       <div>
        {reducerState.stateprop1
           ? <p>stateprop1 is true</p>
           : <p>stateprop1 is false</p>}
       </div>
       <button onClick={dispatchActionSuccess}>
         Dispatch Success
       </button>
    </div>
  )
}

export default TestHookReducer;

```

our reducer: changes stateprop1 to true based on the "SUCCESS" action. 

> ```

Reviewer: I'd suggest changing adding js next to the backticks, so the formatter can highlight the code block:

` ```js`

same in the other code block (use `jsx` there)

import * as ACTIONS from './actions'

export const initialState = {
    stateprop1: false,
}

export const Reducer1 = (state = initialState, action) => {
  switch(action.type) {
    case "SUCCESS":
      return {
        ...state,
        stateprop1: true,
      }
    case "FAILURE":
      return {
        ...state,
        stateprop1: false,
      }
    default:
      console.log(state)
  }
}
```

our actions hold strings of their type. 

```

export const SUCCESS = {
  type: 'SUCCESS'
}

export const FAILURE = {
  type: 'FAILURE'
}

```
Finally our tests: we first test our reducer to make sure it is returning the correct state based on a action. 
Then we test to see if we get the correct output in our JSX based on the reducer state.  

```
import React from 'react';
import ReactDOM from 'react-dom';
import TestHookReducer from '../test_hook_reducer.js';
import {render, fireEvent, cleanup} from '@testing-library/react';
import * as Reducer from '../../store/reducer';
import * as ACTIONS from '../../store/actions';

afterEach(cleanup)

describe('test the reducer and actions', () => {
  it('should return the initial state', () => {

>     expect(Reducer.initialState).toEqual({ stateprop1: false })

Reviewer 2: Since this example is focusing on RTL usage, I think we can drop the unit tests for the reducer in favor of showing how treating the DOM as the unit under test allows the component to be rewritten using state, the class API, redux, etc. without changing the test.

Reviewer: Agree. I'd update the test description text accordingly, too (the "Finally our tests" block)

> Basic example showing how to test useReducer hook. 

Reviewer: ```suggestion
Basic example showing how to test `useReducer` hook.
```

> The component we are testing which changes some text depending on `stateprop1` value.

Reviewer: ```suggestion
The component we are testing, which changes some text depending on `isConfirmed`.
```

```jsx
import React, { useReducer } from 'react'

const initialState = {

>     stateprop1: false,

Reviewer 2: ```suggestion
    isConfirmed: false,
```

>     isConfirmed: false,

Reviewer: ```suggestion
  isConfirmed: false,
```

}

> const Reducer = (state = initialState, action) => {

Reviewer 2: ```suggestion
function reducer(state = initialState, action) {
```

  switch(action.type) {
    case "SUCCESS":
      return {
        ...state,

>         stateprop1: true,

Reviewer 2: ```suggestion
        isConfirmed: true,
```

      }

>     case "FAILURE":

Reviewer: If the "FAILURE" state isn't being used, I'd remove it from the demo, so code is as focused on the topic as it can be :)

      return {
        ...state,

>         stateprop1: false,

Reviewer 2: Not sure if we need a failure case in the example if it is not used in the component, but if so:

```suggestion
        isConfirmed: true,
```

      }
    default:
      console.log(state)
  }
}

> const TestHookReducer = () => {

Reviewer 2: ```suggestion
const Example = () => {
```

>   const [reducerState, dispatch] = useReducer(Reducer, initialState)

Reviewer 2: ```suggestion
  const [state, dispatch] = useReducer(reducer, initialState)
```

  const dispatchActionSuccess = () => {
    dispatch({ type: 'SUCCESS' })
  }

  const dispatchActionFailure = () => {
    dispatch({ type: 'FAILURE' })
  }

  return (
    <div>

>        <div>

Reviewer: the wrapper `<div>` feels unnecessary:

```jsx
<p>
  {state.isConfirmed
    ? 'Confirmed!'
    : 'Waiting for confirmation...'}
</p> 
```

(sry I can't provide multiline suggestions 😞)

>         {reducerState.stateprop1

Reviewer 2: ```suggestion
        {state.isConfirmed
```

>            ? <p>stateprop1 is true</p>

Reviewer 2: ```suggestion
           ? <p>Confirmed!</p>
```

>            : <p>stateprop1 is false</p>}

Reviewer 2: ```suggestion
           : <p>Waiting for confirmation...</p>}
```

       </div>

>  <button onClick={dispatch({type: 'SUCCESS'})}>

Reviewer: ```suggestion
        <button onClick={dispatch({type: 'SUCCESS'})}>
```

>        <button onClick={dispatchActionSuccess}>

Reviewer 2: ```suggestion
 <button onClick={dispatch({type: 'SUCCESS'})}>
```

>          Dispatch Success

Reviewer 2: ```suggestion
         Confirm
```

       </button>
    </div>
  )
}

> export default TestHookReducer

Reviewer 2: ```suggestion
export default Example
```

```

Finally our tests: We are testing to see if we get the correct output in our JSX based on the reducer state.  

```js
import React from 'react'
import ReactDOM from 'react-dom'

> import TestHookReducer from '../test_hook_reducer.js'

Reviewer 2: ```suggestion
import Example from '../example.js'
```

import {render, fireEvent, cleanup} from '@testing-library/react'

afterEach(cleanup)

it('shows success message on click', () => {

>    const { getByText } = render(<TestHookReducer />)

Reviewer: ```suggestion
  const { getByText } = render(<TestHookReducer />)
```

> it('Reducer changes stateprop1 from false to true', () => {

Reviewer 2: ```suggestion
it('shows success message on click', () => {
```

>     expect(getByText(/waiting/i).textContent).toBeInTheDocument()

Reviewer: ```suggestion
  expect(getByText(/waiting/i).textContent).toBeInTheDocument()
```

>    fireEvent.click(getByText("Confirm"))

Reviewer: ```suggestion
  fireEvent.click(getByText("Confirm"))
```

>    expect(getByText(/stateprop1 is/i).textContent).toBe("stateprop1 is false")

Reviewer 2: ```suggestion   
    expect(getByText(/waiting/i).textContent).toBeInTheDocument()
```

>    expect(getByText('Confirmed')).toBeInTheDocument()

Reviewer: ```suggestion
  expect(getByText('Confirmed')).toBeInTheDocument()
```

>    fireEvent.click(getByText("Dispatch Success"))

Reviewer 2: ```suggestion
   fireEvent.click(getByText("Confirm"))
```

>    expect(getByText(/stateprop1 is/i).textContent).toBe("stateprop1 is true")

Reviewer 2: ```suggestion
   expect(getByText('Confirmed')).toBeInTheDocument()
```

Basic example showing how to test useReducer hook. 

The component we are testing which changes some text depending on `stateprop1` value.

```jsx
import React, { useReducer } from 'react'

const initialState = {
    stateprop1: false,
}

const Reducer = (state = initialState, action) => {
  switch(action.type) {
    case "SUCCESS":
      return {
        ...state,
        stateprop1: true,
      }
    case "FAILURE":
      return {
        ...state,
        stateprop1: false,
      }
    default:
      console.log(state)
  }
}

const TestHookReducer = () => {
  const [reducerState, dispatch] = useReducer(Reducer, initialState)

  const dispatchActionSuccess = () => {
    dispatch({ type: 'SUCCESS' })
  }

  const dispatchActionFailure = () => {
    dispatch({ type: 'FAILURE' })
  }

  return (
    <div>
       <div>
        {reducerState.stateprop1
           ? <p>stateprop1 is true</p>
           : <p>stateprop1 is false</p>}
       </div>
       <button onClick={dispatchActionSuccess}>
         Dispatch Success
       </button>
    </div>
  )
}

> export default TestHookReducer;

Reviewer: ```suggestion
export default TestHookReducer
```

```

Finally our tests: We are testing to see if we get the correct output in our JSX based on the reducer state.  

```js
import React from 'react'
import ReactDOM from 'react-dom'
import TestHookReducer from '../test_hook_reducer.js'
import {render, fireEvent, cleanup} from '@testing-library/react'

> 

Reviewer: Could you remove some empty lines? Not sure how to create such suggestion with GitHub UI 🙈

afterEach(cleanup)

it('Reducer changes stateprop1 from false to true', () => {

>    const { container, getByText } = render(<TestHookReducer />);

Reviewer: ```suggestion
   const { getByText } = render(<TestHookReducer />)
```
